### PR TITLE
Add N-way N-shot sampling method

### PR DESCRIPTION
**Pull request by shonohs**

### CODE REVIEW

File: vision_datasets/common/data_manifest.py [hunk 1]
```diff
@@ -324,6 +342,62 @@ def sample_subset_by_ratio(self, sampling_ratio):
         sampled_images = [self.images[i] for i in sampled_image_ids]
         return DatasetManifest(sampled_images, self.labelmap, self.data_type)
 
+    def sample_n_way_n_shot(self, n_way, n_shot, random_seed=0):
+        """
+        Sample N-way N-shot dataset. Sampled dataset has n_way classes and each classes has n_shot images.
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py:347]: nit: each class

File: vision_datasets/common/data_manifest.py [hunk 2]
```diff
@@ -324,6 +342,62 @@ def sample_subset_by_ratio(self, sampling_ratio):
...
+        labels = [[self._get_cid(c) for c in image.labels] for image in self.images]
+
+        # Get lists of images per class_ids.
+        images_by_cids = collections.defaultdict(list)
+        for i, label in enumerate(labels):
+            for c in set(label):
+                images_by_cids[c].append(i)
+
+        # Get a list of classes that have at least N images, and sample N-way classes.
+        sampled_classes = list(rng.sample([c for c in images_by_cids if len(images_by_cids[c]) >= n_shot], n_way))
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py:372]: what is no classes or < N_way classes have images >= n_shot?

File: vision_datasets/common/data_manifest.py [hunk 3]
```diff
@@ -324,6 +342,62 @@ def sample_subset_by_ratio(self, sampling_ratio):
...
+        Each image will have only 1 class. If the original image has multiple classes, annotations for the other classes will be removed.
+
+        This method can throw a RuntimeError if it failed to generate N-way N-shot dataset.
+
+        Args:
+            n_way (int): The number of classes to sample.
+            n_shot (int): The number of images for each classes.
+
+        Returns:
+            A sampled dataset (DatasetManifest)
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py:357]: Can we mention that an exception will be thrown if sampling failed?

File: vision_datasets/common/data_manifest.py (hunk 2 above)
Comment (shonohs) [reply vision_datasets/common/data_manifest.py:372]: Good point. will add a check

File: vision_datasets/common/data_manifest.py [hunk 4]
```diff
@@ -324,6 +342,62 @@ def sample_subset_by_ratio(self, sampling_ratio):
...
+            num_cids = [len(c for c in x if c == c_id) for x in labels]  # [(image_id, num_target_classes)]
+
+            # Get a list of images that has c_id class without other sampled classes.
+            valid_images = list(set(images_by_cids[c_id]) - set([x for key in images_by_cids if key != c_id for x in images_by_cids[key]]))
+            rng.shuffle(valid_images)
+
+            # Sample N-shot images from the list of images.
+            num_samples = 0
+            for image_id in valid_images:
+                if num_cids[image_id] > num_samples - n_shot:  # If the next image has too many samples, skip it.
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py:385]: is num_samples - n_shot < 0 most of times?

File: vision_datasets/common/data_manifest.py (hunk 3 above)
Comment (shonohs) [reply vision_datasets/common/data_manifest.py:357]: sure, will move the line 350 to here.

File: vision_datasets/common/data_manifest.py (hunk 4 above)
Comment (shonohs) [reply vision_datasets/common/data_manifest.py:385]: true lol

File: vision_datasets/common/data_manifest.py [hunk 5]
```diff
@@ -324,6 +342,62 @@ def sample_subset_by_ratio(self, sampling_ratio):
...
+        for i, label in enumerate(labels):
+            for c in set(label):
+                images_by_cids[c].append(i)
+
+        # Get a list of classes that have at least N images, and sample N-way classes.
+        sampled_classes = list(rng.sample([c for c in images_by_cids if len(images_by_cids[c]) >= n_shot], n_way))
+
+        sampled_images = []
+        for c_id in sampled_classes:
+            num_cids = [len(c for c in x if c == c_id) for x in labels]  # [(image_id, num_target_classes)]
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py:376]: I got confused by this part 
len(c for c in x if c == c_id)

Comment (pj-ms) [reply vision_datasets/common/data_manifest.py:376]: is it supposed to be len([c for c in x if c == c_id])? and it only applies to OD cases?

Comment (shonohs) [reply vision_datasets/common/data_manifest.py:376]: yes and no. labels is the result of _get_cid() call.

Comment (pj-ms) [reply vision_datasets/common/data_manifest.py:376]: but why we would like to filter out an image for class i, when image has a lot of bboxes of class i

Comment (shonohs) [reply vision_datasets/common/data_manifest.py:376]: and the comment is old. will fix it

**Comment by shonohs** [comment]
we are changing the design.

**Pull request closed without merging**